hashes/src/lib.rs

```diff
@@ -217,6 +211,30 @@ pub trait Hash:
 
     /// Constructs a hash from the underlying byte array.
     fn from_byte_array(bytes: Self::Bytes) -> Self;
+}
+
+/// Trait which applies to general-purpose hashes that may hash arbitrary data.
+pub trait RawHash: Hash {
```

Comment:
Perhaps we should put `all_zeros` on `Hash` not on `RawHash` since it is only used for Bitcoin specific hashes which implement `Hash`?

Reply:
Only some hashes actually use all-zero values (besides maybe tests). From what I remember, `TxId` (in coinbase prev outs) and `BlockHash` (in genesis block). Only having it on hashes which actually do so is best IMO.

I will try to check how disruptive would it be to remove it from the trait completely.

Reply:
Even better. We could even add an `AllZeros` trait.

Reply:
Unless an algorithm uses `all_zeros` I don't want any more traits. I actually want `RawHash` to be duplicated as inherent methods on all our raw hashes since it's very annoying to have to import the trait. (We need `RawHash` because of merkle trees.)